### ISSUE TITLE
Custom Controller OpenAPI Documentation

### ISSUE THREAD

**Issue by matjones** [mevdschee/php-crud-api#968]
I had an expectation that if I created a custom controller, it would automatically get picked up in OpenAPI documentation. However I don't see my custom route listed when viewing the openapi route. Is there something I'm missing, or do they just not get documented. 

Here is an example of my route:

```
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Tqdev\PhpCrudApi\Cache\Cache;
use Tqdev\PhpCrudApi\Column\ReflectionService;
use Tqdev\PhpCrudApi\Controller\Responder;
use Tqdev\PhpCrudApi\Database\GenericDB;
use Tqdev\PhpCrudApi\Middleware\Router\Router;

class CategoryDropdownController
{

	private $responder;

	public function __construct(Router $router, Responder $responder, GenericDB $db, ReflectionService $reflection, Cache $cache)
	{
		$router->register('GET', '/categorylist', array($this, 'getMyCustomQuery'));
		$this->responder = $responder;
		$this->db = $db;
	}

	public function getMyCustomQuery(ServerRequestInterface $request): ResponseInterface
	{
		$sql = "SELECT * from `categories`"; // Query is simplified for this example
		$pdo_statement = $this->db->pdo()->query($sql);
		$result = $pdo_statement->fetchAll();
		return $this->responder->success(['results' => $result]);
	}
}
```

**Comment by mevdschee** [comment]
>Is there something I'm missing, or do they just not get documented.

No, Im sorry, customer routes are not automatically documented.

**Comment by matjones** [comment]
> No, I'm sorry, customer routes are not automatically documented. Do anyone know how this documentation could be automated? I think it would be a great feature. I'm not sure it is feasible though..

Thanks for the speedy response. It would be great if there was a way to achieve this too.

**Comment by mevdschee** [comment]
Coming back to this after the openapi work in #1078, where @clysss asked the same
thing.

There is no automatic documentation of a custom route, but there is a hook for
writing one by hand, and it has been there since the custom controllers
themselves were added in 43f4000: the `customOpenApiBuilders` setting. It takes a
comma separated list of class names, `OpenApiBuilder` constructs each of them
with the `OpenApiDefinition` and the `ReflectionService` and calls `build()` on
them after the built in builders have run. Such a builder can write anything into
the document, `components|securitySchemes` and `security` included, which is the
other half of what @clysss asked.

The problem is that nobody can find it. The setting is not in the list of
settings in the readme, not in the custom controllers section, and there is no
example of what a builder looks like. I pointed at it in #766 and then never
wrote it down. That is the first thing to fix here.

Beyond the documentation:

- A custom builder is handed the definition and the reflection service and
  nothing else, so it cannot ask `OpenApiMiddlewares` which middlewares are
  enabled, or read a setting of its own. A builder that wants to describe the
  same error responses and the same security as the rest of the document has to
  guess.
- `setSecurity()` runs before the custom builders, which is the right order, but
  `setComponentParameters()` and `setComponentErrors()` run after them and
  overwrite whatever a builder wrote at those keys.
- There is no test that runs a custom builder.

**Comment by mevdschee** [comment]
The list above is done in #1087, along with the documentation that was the first
thing on it.

- The parameters. A builder now gets `OpenApiMiddlewares` next to the definition
  and the reflection service, so it can take the common parameters and the
  status codes from the same place the built in builders take them instead of
  guessing at them. It is appended rather than inserted, so a builder written
  against the old two argument constructor keeps working, as PHP ignores the
  arguments that it does not declare.
- The ordering. `setComponentParameters()` and `setComponentErrors()` now run
  before the custom builders instead of after, so a builder has the last word on
  the whole document, the way `setSecurity()` already let it. Neither of the two
  depends on the built in builders having run, and the generated document is
  byte identical for a config without custom builders.
- The test. The test config registers two builders, one with each constructor,
  so the openapi fixtures cover the new parameters, the old signature and the
  ordering, and the document lint covers what a custom builder writes.

One part of the first item was left out of that PR. Reading a setting of its own
did not work, because `Config::__construct` rejected a dotted key whose prefix
was not an enabled middleware, and a custom controller had the same problem.
That is fixed separately in #1088: the prefix is the short name of the class, so
a class listed as `\App\OpenApi\HelloBuilder` reads the keys that start with
`HelloBuilder`, and what the class receives is a `CustomSettings` bound to its
own prefix rather than the config.

**Comment by clysss** [comment]
@mevdschee  thanks for your  updates about this . 
I've probably to reread all your stuff and the code to understood all the concepts. In the meantime, if it can helps you..
I manipulated severall customControllers to create secured API for more than a year with success so my 2 cents comments :
- if we're able to create customControllers, it's for having a lower access to php-crud-api => in this case, I think that it's not an issue to have to manually create the OpenApiDefinition and so on during the builder. With the help of AI this kind of stupid tasks are just easy to create/maintain. Some of my API are using token, POST and/or GET commands (including for example php-crud-api filters concept), and even cookies or specific header. So I perfectly agree that I've to precise all this manually to the OpenAPI. By doing that, I'll also improve the documentation of my project (and the automatisation of it), so I'll never considerate this as lost time...
- concerning securitySettings, In my opinion, customControllers aren't supposed to have access to all "knowledge" of other middleware properties directly. They could for example try something (and receive an error if this element is refused/inexistant/blocked) but not make discovering of the environnement. Perhaps middleware should have this capabilities, but I've not enough play with them to be able to have an opinion about this. So, If a custom controler can just add in the setSecurity an item, it is perfect. If not, if there is an option to define the setSecurity during config (manualy), It should be enough.
I publish severall month ago a customController here for documentation : I could try to update it to follow the new openAPI documentation ?

**Comment by mevdschee** [comment]
@clysss thanks, that is useful, and one part of it changed the PR.

On writing the definition by hand: agreed, that stays the design. #1087 only
makes the hook findable and shows what a builder looks like. Nothing in it
pushes you off the way you work now, because the extra constructor parameter is
appended, so a builder that declares only the definition and the reflection
service keeps working. There is a test for exactly that.

On not discovering the environment: there is no boundary to defend there, a
builder is a class in the same process that already gets the `ReflectionService`
and can do whatever PHP can do. But you are right in a way that mattered. I had
also handed the builder the whole `Config`, which carries `getUsername()` and
`getPassword()`, and a class that writes documentation has no use for those. It
was there so a builder could read a setting of its own, and it turns out it
cannot anyway, because `Config` rejects a dotted key whose prefix is not an
enabled middleware. So it is out again. A builder gets `OpenApiMiddlewares` and
nothing more, which answers only what the enabled middlewares mean for the
document: the parameters that every operation has because a middleware reads
them, and the status codes a middleware can return before the end-point is
reached.

On the security scheme, both of the things you asked for already work.

A builder can add one:

```php
$this->openapi->set('components|securitySchemes|tokenAuth|type', 'apiKey');
$this->openapi->set('components|securitySchemes|tokenAuth|in', 'header');
$this->openapi->set('components|securitySchemes|tokenAuth|name', 'X-Token');
$this->openapi->set('security||tokenAuth', []);
```

And your fallback, declaring it in the config, works too and needs no class at
all. "openApiBase" is the JSON that the document is built on top of, so it can
carry "components" and "security" next to "info":

```php
'openApiBase' => '{"info":{"title":"My API","version":"1.0.0"},"components":{"securitySchemes":{"tokenAuth":{"type":"apiKey","in":"header","name":"X-Token"}}},"security":[{"tokenAuth":[]}]}',
```

`setSecurity()` appends per name rather than replacing, so what the enabled
authentication middlewares describe lands next to yours instead of wiping it.
With "dbAuth" on and the above in the config, the document has both:

```
schemes  ["tokenAuth","dbAuth"]
security [{"tokenAuth":[]},{"dbAuth":[]},[]]
```

That was written down nowhere, so it is in the readme now.

Please do update your example, a year of real use makes it a better one than
mine. On the constructor: the two parameter one keeps working and always will,
so there is no rush. #1087 appends `OpenApiMiddlewares` as a third parameter and
#1088 appends a `CustomSettings` as a fourth, which is the piece that lets a
builder or a controller read a setting of its own. Do not write against the
`Config` parameter that an earlier comment of mine mentioned, that never
shipped.

**Comment by mevdschee** [comment]
Closing summary, as the comments above were written while this was still moving.

A custom route is still not documented on its own, and is not going to be. The
router knows the method and the path, nothing about the parameters the
end-point reads or the document it answers with, so there is nothing to
generate a description from. What there is, and what was missing from the
readme all along, is a hook to write it by hand: the `customOpenApiBuilders`
setting. It is in the list of settings now, and the custom controller section
has a "Documenting a custom controller" part with an example.

A builder is constructed with the definition, the reflection service, an
`OpenApiMiddlewares` and a `CustomSettings`. The middlewares answer what the
enabled middlewares mean for the document, so a hand written operation gets the
same common parameters and the same error responses as the rest of it instead
of guessing. The settings are the config keys that start with the short name of
the class. Both are appended, so a builder written against the original two
parameter constructor keeps working, and there is a test for that.

A builder can write anywhere in the document, `components|securitySchemes` and
`security` included, and it now runs after everything built in, so it has the
last word. A scheme that is the same on every request does not need a builder
at all, "openApiBase" carries it, and what the enabled authentication
middlewares describe is added next to it rather than in place of it.

Custom controllers got the same settings parameter, since they had the same
gap.

Shipped in #1087 and #1088.

**Comment by mevdschee** [comment]
Released in v2.16.6: Improve customOpenApiBuilders

**Comment by clysss** [comment]
Thanks again @mevdschee  : I'll try to update the example on #1037 asap (or something similar to...)
